Source/WebCore/Modules/reporting/ReportingObserver.cpp

```diff
@@ -53,15 +80,63 @@ ReportingObserver::~ReportingObserver()
 
 void ReportingObserver::disconnect()
 {
+    // https://www.w3.org/TR/reporting-1/#dom-reportingobserver-disconnect
+    if (m_reportingScope)
+        m_reportingScope->unregisterReportingObserver(*this);
 }
 
 void ReportingObserver::observe()
 {
+    ASSERT(m_reportingScope);
+    if (!m_reportingScope)
+        return;
+
+    // https://www.w3.org/TR/reporting-1/#dom-reportingobserver-observe
+    m_reportingScope->registerReportingObserver(*this);
+
+    if (!m_options.buffered)
+        return;
+
+    m_options.buffered = false;
+
+    // For each report in global’s report buffer, queue a task to execute § 4.3 Add report to observer with report and the context object.
+    m_reportingScope->appendQueuedReportForRelevantType(*this);
 }
 
-Vector<Report> ReportingObserver::takeRecords()
+auto ReportingObserver::takeRecords() -> Vector<Ref<Report>>
 {
-    return { };
+    // https://www.w3.org/TR/reporting-1/#dom-reportingobserver-takerecords
+    return WTFMove(m_queuedReports);
+}
+
+void ReportingObserver::appendQueuedReportIfCorrectType(const Ref<Report>& report)
+{
+    // https://www.w3.org/TR/reporting-1/#add-report
+    // Step 4.3.1
+    if (!isVisibleToReportingObservers(report->type()))
+        return;
+    
+    // Step 4.3.2
+    if (m_options.types && m_options.types->contains(report->type()))
+        return;
+    
+    // Step 4.3.3:
+    m_queuedReports.append(report);
+    
+    auto reportsForCallback = takeRecords();
```

Comment:
This is not right. Taking the reports must happen inside the queued task, not beforehand like this.

Reply:
I see. Fixed.